vms/avm/states/state.go

```diff
@@ -576,3 +586,158 @@ func (s *state) writeStatuses() error {
+	}
+
+	i := 1
+	for statusIter.Next() {
+		i++
+
+		txIDBytes := statusIter.Key()
+		statusBytes := statusIter.Value()
+		if err := s.cleanupTx(lock, txIDBytes, statusBytes, txIter); err != nil {
+			return err
+		}
+		if i%pruneCommitLimit == 0 {
+			// We must hold the lock during committing to make sure we don't
+			// attempt to commit to disk while a block is concurrently being
+			// accepted.
+			lock.Lock()
+			err := s.Commit()
+			lock.Unlock()
+			if err != nil {
+				return err
+			}
+
+			log.Info("committing state pruning",
+				zap.Int("count", i),
+			)
+		}
+	}
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	errs := wrappers.Errs{}
+	errs.Add(
+		s.Commit(),
+		statusIter.Error(),
+		txIter.Error(),
+	)
+
+	// Make sure we flush the original caches before re-enabling them to prevent
+	// surfacing any stale data.
+	oldStatusCache.Flush()
+	oldTxCache.Flush()
+	s.statusCache = oldStatusCache
+	s.txCache = oldTxCache
+
+	log.Info("finished state pruning",
+		zap.Int("count", i),
+	)
+
+	return errs.Err
+}
+
+func (s *state) cleanupTx(lock sync.Locker, txIDBytes []byte, statusBytes []byte, txIter database.Iterator) error {
+	// After the linearization, we write txs to disk without statuses to mark
+	// them as accepted. This means that there may be more txs than statuses and
+	// we need to skip over them.
+	//
+	// Note: We do not need to remove UTXOs consumed after the linearization, as
+	// those UTXOs are guaranteed to have already been deleted.
+	if err := skipTo(txIter, txIDBytes); err != nil {
+		return err
+	}
+
+	statusInt, err := database.ParseUInt32(statusBytes)
+	if err != nil {
+		return err
+	}
+	status := choices.Status(statusInt)
+
+	if status == choices.Accepted {
+		txBytes := txIter.Value()
+		tx, err := s.parser.ParseGenesisTx(txBytes)
+		if err != nil {
+			return err
+		}
+
+		utxos := tx.Unsigned.InputUTXOs()
+
+		// Locking is done here to make sure that any concurrent verification is
+		// performed with a valid view of the state.
+		lock.Lock()
+		defer lock.Unlock()
+
+		// Remove all the UTXOs consumed by the accepted tx. Technically we only
+		// need to remove UTXOs consumed by operations, but it's easy to just
+		// remove all of them.
+		for _, UTXO := range utxos {
+			if err := s.utxoState.DeleteUTXO(UTXO.InputID()); err != nil {
+				return err
+			}
+		}
+	} else {
+		lock.Lock()
+		defer lock.Unlock()
+
+		// This tx wasn't accepted, so we can remove it entirely from disk.
+		if err := s.txDB.Delete(txIDBytes); err != nil {
+			return err
+		}
+	}
+	// By removing the status, we will treat the tx as accepted if it is still
+	// on disk.
+	return s.statusDB.Delete(txIDBytes)
+}
+
+// skipTo advances [iter] until its key is equal to [targetKey]. If [iter] does
+// not contain [targetKey] an error will be returned.
+//
+// Note: [iter] will always have [Next()] called at least once.
+func skipTo(iter database.Iterator, targetKey []byte) error {
+	for {
+		if !iter.Next() {
+			return database.ErrNotFound
```

Comment:
maybe we should include in the error the `targetKey` we're looking for? Would help for debugging. Same for the `case -1` below.